src/Tasks/Hash.cs

```diff
@@ -49,7 +54,11 @@ public override bool Execute()
                     {
                         foreach (var item in ItemsToHash)
                         {
-                            stringBuilder.Append(item.ItemSpec);
+                            foreach (char c in item.ItemSpec)
```

Comment:
If `IgnoreCase` is `false`, would it be faster to just have the `StringBuilder` append `item.ItemSpec` instead of each character since it uses pointers and strcpy?
http://index/?query=stringbuilder.app&rightProject=mscorlib&file=system%5Ctext%5Cstringbuilder.cs&line=658

Reply:
I tried various methods:

Method | Mean | Error | StdDev | Gen 0 | Gen 1 | Allocated
-- | -- | -- | -- | -- | -- | --
StringBuilderAppend | 12.60 us | 0.0173 us | 0.0144 us | 15.6097 | 2.2278 | 96.45 KB
StringBuilderAppendIgnoreCase | 77.81 us | 0.5760 us | 0.5388 us | 24.9023 | 3.5400 | 154.33 KB
StringBuilderAppendEachChar | 59.33 us | 1.0765 us | 1.0070 us | 15.5640 | 2.1973 | 96.45 KB
StringBuilderAppendEachCharIgnoreCase | 506.68 us | 0.9487 us | 0.8410 us | 14.6484 | 1.9531 | 96.45 KB


IgnoreCase in this is just calling `ToUpperInvariant()` on the string or the character. I did a *.cs list in Microsoft.Build project for the input and pre-calculated the size. I think we should just do `Append()` when false and `ToUpperInvariant()` on the string when true, even though it uses a bit more memory.

Reply:
Why ToUpper? The data is paths, which tend to be lowercase, so doing ToLower would involve fewer case changes.

Reply:
Perf is about the same. I've always used ToUpper for things like this or comparison. As I understand it (but have never directly experienced) there are languages where it is better to use ToUpper. There used to be an FxCop or StyleCop rule that would suggest always using ToUpper I believe.

Method | Mean | Error | StdDev | Gen 0 | Gen 1 | Allocated
-- | -- | -- | -- | -- | -- | --
StringBuilderAppendIgnoreCaseToUpper | 77.68 us | 0.2029 us | 0.1898 us | 24.9023 | 3.5400 | 154.33 KB
StringBuilderAppendIgnoreCaseToLower | 76.91 us | 0.1107 us | 0.0982 us | 24.9023 | 3.5400 | 154.33 KB